homeassistant/components/proxmoxve/__init__.py

```diff
@@ -200,56 +202,65 @@ async def async_update_data() -> dict[str, Any] | None:
 
         def poll_api() -> dict[str, Any] | None:
             """Call the api."""
-            vm_status = call_api_container_vm(proxmox, node_name, vm_id, vm_type)
-            return vm_status
+            api_status = {}
 
-        vm_status = await hass.async_add_executor_job(poll_api)
-
-        if vm_status is None:
-            _LOGGER.warning(
+            try:
+                if api_category == ProxmoxType.QEMU:
+                    api_status = (
+                        proxmox.nodes(node_name).qemu(vm_id).status.current.get()
+                    )
+                elif api_category == ProxmoxType.LXC:
+                    api_status = (
+                        proxmox.nodes(node_name).lxc(vm_id).status.current.get()
+                    )
+            except AuthenticationError as error:
+                raise ConfigEntryAuthFailed from error
+            except SSLError as error:
+                raise UpdateFailed from error
+            except ConnectTimeout as error:
+                raise UpdateFailed from error
+            except ResourceException as error:
+                raise UpdateFailed from error
+
+            LOGGER.debug("API Response: %s", api_status)
+            return api_status
+
+        api_status = await hass.async_add_executor_job(poll_api)
+
+        if api_status is None:
+            LOGGER.warning(
                 "Vm/Container %s unable to be found in node %s", vm_id, node_name
             )
             return None
 
-        return parse_api_container_vm(vm_status)
+        return parse_api_proxmox(api_status, api_category)
```

Comment:
`parse_api_proxmox` doesn't do much - I don't think it needs its own function.
I also think you should be returning a dataclass, not a dict.

```suggestion
        return ProxmoxData(status=api_status["status"], name=api_status["name"])
```

Reply:
Currently it does little, but the idea is to use it to implement more information for sensors of VMs and nodes.

Some information will be evaluated and treated before, depending on the type of promox resource.

An example:

```
def parse_api_proxmox(
    status: dict[str, Any],
    api_category: ProxmoxType,
) -> dict[str, Any] | None:
    """Get the container or vm api data and return it formatted in a dictionary.

    It is implemented in this way to allow for more data to be added for sensors
    in the future.
    """

    memory_free: float | None = None

    LOGGER.debug("Raw status %s: %s", api_category, status)

    if api_category == ProxmoxType.Proxmox:
        return {
            ProxmoxKeyAPIParse.STATUS: status[ProxmoxKeyAPIParse.STATUS],
            ProxmoxKeyAPIParse.NAME: status[ProxmoxKeyAPIParse.NAME],
        }

    if api_category is ProxmoxType.Node:
        return {
            ProxmoxKeyAPIParse.STATUS: status[ProxmoxKeyAPIParse.STATUS],
            ProxmoxKeyAPIParse.UPTIME: status[ProxmoxKeyAPIParse.UPTIME],
            ProxmoxKeyAPIParse.MODEL: status[ProxmoxKeyAPIParse.CPU_INFO][
                ProxmoxKeyAPIParse.MODEL
            ],
            ProxmoxKeyAPIParse.CPU: status[ProxmoxKeyAPIParse.CPU],
            ProxmoxKeyAPIParse.MEMORY_USED: status[ProxmoxKeyAPIParse.MEMORY][
                ProxmoxKeyAPIParse.USED
            ],
            ProxmoxKeyAPIParse.MEMORY_TOTAL: status[ProxmoxKeyAPIParse.MEMORY][
                ProxmoxKeyAPIParse.TOTAL
            ],
            ProxmoxKeyAPIParse.MEMORY_FREE: status[ProxmoxKeyAPIParse.MEMORY][
                ProxmoxKeyAPIParse.FREE
            ],
            ProxmoxKeyAPIParse.SWAP_TOTAL: status[ProxmoxKeyAPIParse.SWAP][
                ProxmoxKeyAPIParse.TOTAL
            ],
            ProxmoxKeyAPIParse.SWAP_FREE: status[ProxmoxKeyAPIParse.SWAP][
                ProxmoxKeyAPIParse.FREE
            ],
            ProxmoxKeyAPIParse.DISK_USED: status[ProxmoxKeyAPIParse.DISK],
            ProxmoxKeyAPIParse.DISK_TOTAL: status[ProxmoxKeyAPIParse.MAX_DISK],
        }

    if api_category in (ProxmoxType.QEMU, ProxmoxType.LXC):
        if ProxmoxKeyAPIParse.FREE_MEM in status:
            memory_free = status[ProxmoxKeyAPIParse.FREE_MEM]
        else:
            memory_free = (
                status[ProxmoxKeyAPIParse.MAX_MEM] - status[ProxmoxKeyAPIParse.MEM]
            )

        health = None
        if ProxmoxKeyAPIParse.QMP_STATUS in status:
            health = status[ProxmoxKeyAPIParse.QMP_STATUS]

        return {
            ProxmoxKeyAPIParse.STATUS: status[ProxmoxKeyAPIParse.STATUS],
            ProxmoxKeyAPIParse.HEALTH: health,
            ProxmoxKeyAPIParse.UPTIME: status[ProxmoxKeyAPIParse.UPTIME],
            ProxmoxKeyAPIParse.NAME: status[ProxmoxKeyAPIParse.NAME],
            ProxmoxKeyAPIParse.CPU: status[ProxmoxKeyAPIParse.CPU],
            ProxmoxKeyAPIParse.MEMORY_TOTAL: status[ProxmoxKeyAPIParse.MAX_MEM],
            ProxmoxKeyAPIParse.MEMORY_USED: status[ProxmoxKeyAPIParse.MEM],
            ProxmoxKeyAPIParse.MEMORY_FREE: memory_free,
            ProxmoxKeyAPIParse.NETWORK_IN: status[ProxmoxKeyAPIParse.NET_IN],
            ProxmoxKeyAPIParse.NETWORK_OUT: status[ProxmoxKeyAPIParse.NET_OUT],
            ProxmoxKeyAPIParse.DISK_TOTAL: status[ProxmoxKeyAPIParse.MAX_DISK],
            ProxmoxKeyAPIParse.DISK_USED: status[ProxmoxKeyAPIParse.DISK],
        }

    status = None
```

Reply:
I would still use dataclasses instead of a dict.
I would also use separate typed coordinator (see `sfr_box` or `renault` for sample generic coordinators)

Reply:
> I would also use separate typed coordinator (see `sfr_box` or `renault` for sample generic coordinators)

This I still need to understand better to do...